openedx/stanford/djangoapps/course_utils/management/commands/course_grep.py

```diff
@@ -0,0 +1,108 @@
+"""
+Search component data for specific text
+"""
+from django.core.management.base import BaseCommand
+
+from opaque_keys.edx.keys import CourseKey
+from xmodule.capa_module import CapaDescriptor
+from xmodule.html_module import HtmlBlock
+from xmodule.modulestore.django import modulestore
+
+_supported_component_types = [
+    'capa',
+    'html',
+]
+
+class Command(BaseCommand):
+    """
+    Search component data for specific text
+    """
+    help = __doc__
+
+    def add_arguments(self, parser):
+        parser.add_argument(
+            '-t',
+            '--type',
+            action='append',
+            choices=_supported_component_types,
+            help=(
+                'Limit search to specific component types. '
+                'By default, search all component types. '
+            ),
+        )
+        parser.add_argument(
+            '-c',
+            '--course',
+            action='append',
+            help=(
+                'Limit search to a specific course. '
+                'By default, search all courses. '
+            ),
+        )
+        parser.add_argument(
+            'search_terms',
+            help=(
+                'Specify one or more search terms, '
+                'case-sensitive. '
+            ),
+            nargs='*',
+        )
+
+    def handle(self, *args, **options):
+        """
+        Execute the command
+        """
+        terms = options['search_terms'] or ['',]
+        component_types = options.get('type', _supported_component_types)
+        course_ids = options.get('course')
+        results = search(terms, component_types, course_ids)
+        for course, component in results:
+            print(unicode(course.id), component.data)
```

Comment:
What would the output of this command look like?

Generally speaking, we're looking for components in order to go and modify them. In this case, the url to the component in studio is the most useful thing.

Reply:
Yeah, this isn't optimal; just my first draft.
The output looks like:

```
('ORG/COURSE/RUN', 'CONTENTS OF COMPONENT HERE')
```

I think having both of these pieces of data is helpful:
- course_id makes it easy to see, at a glance, which course this occurs in
- data is helpful for the context and determining how much/what kind of match it is

That said, I agree; URL would be very helpful, so we should add that too.
I think I omitted it just for expediency; I didn't know immediately how to construct the URL.

Do you know of a helper method for that?
If not, do you know how these are constructed?

Maybe the output can be a tab-delimited list of:
```
course_id        https://.......        lorem ipsum etc etc
```

Tab-delimiting lets us pipe this output to `awk`, `cut`, etc. for additional command-line processing, like
```sh
# This will give you a list of all URLs
./manage.py cms --settings=devstack course_grep foobar | awk '{print $2}'

# This will give you a list of all impacted courses
./manage.py cms --settings=devstack course_grep foobar | awk '{print $1} | sort | uniq'
```

Reply:
That would be of the form:

https://studio.<instance>/container/block-v1:<triplet>+type@problem+block@<module_id>

module_id is the block_id of the structure.